### DISCUSSION TITLE
Does the following snippet from typeshed correctly and completely type `classmethod`, and if so, does a static type checker need to special case (hard code) `classmethod` instead of treating it correctly like a decorator?

### DISCUSSION THREAD

**Discussion by kwsp** [python/typing#1459] [Polls]
https://github.com/python/typeshed/blob/f28cb8b8562ccc382d018129ba4886f241c6db9c/stdlib/builtins.pyi#L146-L160

```python
class classmethod(Generic[_T, _P, _R_co]):
    @property
    def __func__(self) -> Callable[Concatenate[type[_T], _P], _R_co]: ...
    @property
    def __isabstractmethod__(self) -> bool: ...
    def __init__(self, __f: Callable[Concatenate[type[_T], _P], _R_co]) -> None: ...
    @overload
    def __get__(self, __instance: _T, __owner: type[_T] | None = None) -> Callable[_P, _R_co]: ...
    @overload
    def __get__(self, __instance: None, __owner: type[_T]) -> Callable[_P, _R_co]: ...
    if sys.version_info >= (3, 10):
        __name__: str
        __qualname__: str
        @property
        def __wrapped__(self) -> Callable[Concatenate[type[_T], _P], _R_co]: ...
```

[Pyright special cases (hard codes) `classmethod`](https://github.com/microsoft/pyright/blob/900fbd865ebd9ed9156d12d91bd95f3e5376b5cf/packages/pyright-internal/src/analyzer/decorators.ts#L228-L253), and I'm pretty sure other type checkers do too. By doing so, they avoid having to parse the above type annotations and work with descriptors correctly.

**Comment by erictraut** [comment]
I think your question is based on faulty assumptions. Refer to [this explanation](https://github.com/microsoft/pyright/issues/5865#issuecomment-1701910400) for details.

**Reply by kwsp** [reply]
@erictraut Thanks for taking the time to explain! I will need some time to digest this information.

**Comment by kwsp** [comment]
@erictraut I made a [proposal for `typing.no_discard`](https://discuss.python.org/t/proposal-typing-no-discard-a-decorator-to-indicate-that-the-return-value-should-not-be-discarded/33138). Would love to get your input!
